### ISSUE TITLE
Enhancement/design idea: encrypt everything: client-side end-to-end encryption using OpenPGP JavaScript implementation, not requiring additional software to be installed [$25]

### ISSUE THREAD

**Issue by Wikinaut** [owncloud/core#106]
I file this as suggestion to overcome several problems with the server-side encryption of pre-4.5 versions.

Situation ownCloud (as far as I understand it)
- when using the server-side encryption, you cannot share files. Files are encrypted per single key for all files in the user directory. Sharing would require to pass the key to the person with whom you want to share the file.

Current developments:
- http://sebsauvage.net/wiki/doku.php?id=php:zerobin is a opensource online pastebin/discussion board where the server has zero knowledge of hosted data. Data is encrypted/decrypted in the browser using 256 bits AES. See also http://www.heise.de/security/meldung/Interessantes-Konzept-fuer-verschluesselte-Pastes-1541988.html
- http://gpg4browsers.recurity.com/ and http://www.openpgpjs.org/ (new [1]) and http://www.heise.de/newsticker/meldung/OpenPGP-im-Browser-1381826.html

Using the mentioned techniques it would become possible 
- to have a possibility for a platform independent (browser-) client-side encryption and decryption of files. All data traffic to the server would be encrypted, all files are only stored encrypted on the server, and no server-side encryption will be necessary any longer.
- to assign all users a user key (each user has their own user key), and all their files got encrypted (client-side) with their specific user's key
- to share (encrypted) files, given that they are encrypted with more than the key of the user, for example the user key and a key for another user, or for another group who shares one group key.
- enterprise users, where in enterprises usually the installation of additional software is prohibited, can participate as any other user, because the en/decryption takes place in their browsers and does not require installation of  any software (GPG, openPGP, ...) except a contemporary HTML5 browser which is capable to run the client-side crypto algorithms.
- all traffic is encrypted
- no plain files on the server

[0] https://github.com/openpgpjs/openpgpjs/wiki/Introduction#wiki-impl-js Other OpenPGP JavaScript implementations
[1]  http://www.openpgpjs.org/ 
This project aims to provide an Open Source OpenPGP library in JavaScript so it can be used on virtually every device. Instead of other plugins that are aimed at using e.g. NPAPI (meaning they are intended to run native code), OpenPGP.js is meant to bypass this requirement (i.e. people will not have to install gpg on their machines in order to use the software). The idea is to implement all the needed OpenPGP functionality in a JavaScript library, to reuse it in other projects and to provide example applications and browser plugins. It should allow you to sign, encrypt, decrypt, and verify any kind of text - in particular e-mails - as well as managing keys.

[2] https://github.com/openpgpjs/openpgpjs/issues/12 Feature: PGP encrypted file upload/download
[3] https://github.com/hellais/up-crypt client side AES encrypted file upload.

[UPDATED 20121210]

[4] http://www.mailvelope.com/ Mailvelope is a browser extension that allows to exchange encrypted emails following the OpenPGP encryption standard.
[5] https://github.com/toberndo/mailvelope

[updated 20130102]

[6] http://www.dp-dhl.com/en/media_relations/press_releases/2012/docwallet_documents_manager_for_the_ipad.html Deutsche Post AG end-to-end AES-256 encryption for documents 
[7] http://www.heise.de/ct/inhalt/2013/02/42/ "PDF-Tresor von der Post" article in "c't" 2013/2 page 42 

[updated 201304261]

[8] https://github.com/toberndo/mailvelope/issues/45 regarding private key security

[updated 20131114]

I just found these cloud services claiming to offer end-to-end encrypted cloud services

[9]  SpiderOak https://spideroak.com/
[10] TeamDrive http://www.teamdrive.com/de/

**Comment by jospoortvliet** [comment]
@orion1024  this is cool stuff ;-)

**Comment by orion1024** [comment]
Hello,

an update on things.
As I finished the first document describing intents and features, I moved on to the design of the crypto system.
As of now, it is far from finished, but I think it is at a point where someone that wants to contribute ideas/remarks has enough material to get a good idea of where I'm headed. 

Here is the document  (remember : this is ongoing work !) :
https://cloud.airmail.fr/public.php?service=files&t=d475f3c9d77e99d8b413977729be5f0e

And the first document with intents and features (needs some updates with the release of ownCloud 8). Still open for review of course :
https://cloud.airmail.fr/public.php?service=files&t=a9fe1a55a59374dd6bc399f4ff4d60db

Frank/jos : thanks for the support :)

**Comment by jospoortvliet** [comment]
@LukasReschke I bet you want to have a look ;-)

**Comment by orion1024** [comment]
Hello,

a new update.

The encrypted storage system (at least the core of it)  is stabilized as far as I am concerned, so I switched to making a graph that explains more visually how the system works.
The text document grew a bit too difficult to my taste and I thought a graph would help other people to understand it more easily.

You can get the current graph here : https://cloud.airmail.fr/index.php/s/7g6gd2IbJMHtpJ9
Use yED to read it (free software).
Here is the PDF if you want a quick look at it without needing yED: https://cloud.airmail.fr/index.php/s/hTlnrGca3c6zUiM

What's next :

**Comment by orion1024** [comment]
Hello,

a new update. First, I dubbed the "ownCloud client-side encryption" feature as ownCrypt, for brievity sake. This will do at least as a temporary name.

The encrypted storage system (at least the core of it)  is stabilized as far as I am concerned, so I switched to making a graph that explains more visually how the system works.
The text document grew a bit too difficult to my taste and I thought a graph would help other people to understand it more easily.

You can get the current graph [here] (https://cloud.airmail.fr/index.php/s/7g6gd2IbJMHtpJ9). Use yED to read it (free software).
[Here is the PDF] (https://cloud.airmail.fr/index.php/s/hTlnrGca3c6zUiM) if you want a quick look at it without needing yED.

What's next :
* putting more use cases in graph
* most importantly, I will begin seeking out feedback and advice from others on the cryptographic strength and scalability of the current design before I build further on the current design. As I don't have any prior experience in crypto engineering I need more confidence in the foundation of the cryptosystem.
* begin work on other aspects of the cryptosystem that are **not directly** dependant on how the files are stored : key exchange, key management, for instance. Key exchange particularly is a big and interesting topic. 
* integration with ownCloud current capabilities (for instance, the new sharing API coming in OC 8.1 will be instrumental for ownCrypt).

@schiesbn 
@Raydiation 

I am very interested in your opinion on the current state of the design. Even more so if, according to **your** knowledge of the current server and client code/roadmap, you think that some of my design intents might conflict with yours, or simply not mesh well together.

**Comment by MartinDesigner** [comment]
Regarding only to design ideas, responsive and creativity, look at this stuff:
![imghome](https://cloud.githubusercontent.com/assets/13809960/9289329/0bef602c-436c-11e5-80fc-3ec07f82dd8d.jpg)
So, as a supplement, however as a main point try this amazing free app. Createer.com is an online software for easily designing online ads, responsive landing pages and print materials such as flyers, posters, covers, invitations and more. Createer lets you easily share your designs on social media, send them via email, download as HTML, image or PDF. You can even design UI mockups for mobile as for standard display (html, js, css always generating without any coding required from you). Createer also includes 3,000 predesigned templates (browse here: https://beta2.createer.com/Templates?ghi) and over 250,000 images (illustrations, photos and icons) to work with in its user-friendly drag’n’drop enviroment. And finally this amazing stuff is here: https://beta2.createer.com/?ghi

**Comment by jospoortvliet** [comment]
https://conference.owncloud.org/conference/oCC2015/proposal/72 talk is on YouTube.

**Comment by jospoortvliet** [comment]
https://www.youtube.com/watch?v=EI-Zyf8lOws
https://www.youtube.com/watch?v=5AM3340c9DQ
;-)

Did you make any progress lately, @orion1024 ?

**Comment by orion1024** [comment]
Thanks for the links and the work on the videos Jos !

Right now I'm busy studying the client code and how to best integrate ownCrypt into it, since ownCrypt code will need to run deep in the sync code and we want to avoid my code getting in the way of regular development.
Once I get a good grasp on this I will start working with the desktop guys to get their opinion/approval on my plans.

In parallel, there is also undergoing work with a French university in order to get more people involved in the project, especially people knowledgeable in cryptography. Nothing confirmed yet, hence my silence, but I'm hopeful.
Stay tuned...

**Comment by jospoortvliet** [comment]
/me just occasionally likes to kick @orion1024 to get some status info :innocent:

**Comment by orion1024** [comment]
We said no kicks :)

I've been at nearly full throttle for the past month, so if everything goes as planned you will hear from me around new Eve.
Finger crossed !

**Comment by orion1024** [comment]
Here is the promised update.

I created [a new issue in the client repository] (https://github.com/owncloud/client/issues/4327) to kick off the ownCrypt integration discussion.
First step is to agree with the client team how the integration should happen codewise, and then off we code.

I will keep posting updates here but most of the technical discussion will now take place at the new issue.

**Comment by jospoortvliet** [comment]
Awesome! Let's see what people say. I guess you'll have to wait a bit until they're all back on the job, though...

**Comment by tflidd** [comment]
There is a new client-side transparent encryption solution (open source): https://cryptomator.org/
It should be possible to use this on top of owncloud.

**Comment by fpietrosanti** [comment]
OpenPGP.js as per 13 Feb 2015, do support entire operation via native WebCrypto API https://github.com/openpgpjs/openpgpjs/releases/tag/v2.0.0 .

**Comment by JonasDralle** [comment]
+1
this would be awesome

**Comment by xlyz** [comment]
+1

**Comment by Ardakilic** [comment]
+1 for this, too.

**Comment by Hamsterman** [comment]
+1
